### PR TITLE
Add functional test for the UCS service.

### PR DESCRIPTION
**Pull request by tannoa2**
Note: This PR will fail because the CI infrastructure doesn’t include the virtual ucs VMs.
@keedya  is working on. However, this work has been tested in sandbox and it passes.
Please go ahead and review the PR despite of the failure.
The tests in this file tests:
-The APIs that are provided by the ucs-service:
https://github.com/RackHD/ucs-service 
-The ucs discovery workflow
-ucs catalog workflow

@hohene @gpaulos @stuart-stanley @RackHD/veyron

### CODE REVIEW

**Comment by keedya** [comment]
depends on https://github.com/RackHD/on-build-config/pull/105

**Comment by keedya** [comment]
test this please

**Comment by keedya** [comment]
this PR has a dependency: https://github.com/RackHD/on-build-config/pull/105

**Comment by tannoa2** [comment]
test this please

**Comment by keedya** [comment]
test this please

**Comment by keedya** [comment]
running Jenkins now after updating the jenkinsCi FIT to job and adding ucs service to the testing pool

**Comment by keedya** [comment]
Pre-existing failure; retesting
test this please

File: test/tests/rackhd20/test_rackhd20_api_ucs.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+        ucs_service = self.RACKHD_IP + ":" + self.UCS_PORT + "/"
+        ucs_manager = self.UCS_IP
+        if identifier is None:
+            url = "http://" + ucs_service + api + "?host=" + ucs_manager + " &user=ucspe&password=ucspe"
+        else:
+            url = "http://" + ucs_service + api + "?host=" + ucs_manager + " &user=ucspe&password=ucspe"\
+                  "&identifier=" + identifier
+        return url
+
+    def test_api_20_ucs_log_in(self):
```
Comment (hohene) [inline test/tests/rackhd20/test_rackhd20_api_ucs.py]: Should add a test that if ucs.ip or ucs.port is not defined, the test script should bail out.
test_check_usc_params():
  if 'ucs_ip' not in fitcfg() or if 'ucs_ip' not in fitcfg():
     assert.fail.
Rest of tests, then can add a depends line.
@depends(after=test_check_ucs_params)
that way they get skipped, if the test is run against a stack without ucs.

Or you could sys.exit the script at assert.fail, tho that isn't very clean.

File: test/tests/rackhd20/test_rackhd20_api_ucs.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+    def test_api_20_ucs_get_sys(self):
+        """
+        Test the /sys ucs API
+        :return:
+        """
+        url = self.ucs_url_factory("sys")
+        api_data = fit_common.restful(url)
+        self.assertEqual(api_data['status'], 200,
+                         'Incorrect HTTP return code, expected 201, got:' + str(api_data['status']))
```
Comment (hohene) [inline test/tests/rackhd20/test_rackhd20_api_ucs.py]: change comment, expected 200 as opposed to 201

File: test/tests/rackhd20/test_rackhd20_api_ucs.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+    def test_api_20_ucs_log_in(self):
+        """
+        Test the /logIn ucs API
+        :return:
+        """
+        url = self.ucs_url_factory("login")
+        api_data = fit_common.restful(url)
+        self.assertEqual(api_data['status'], 200,
+                         'Incorrect HTTP return code, expected 201, got:' + str(api_data['status']))
```
Comment (hohene) [inline test/tests/rackhd20/test_rackhd20_api_ucs.py]: Change comment, expected 200 as opposed to 201.

File: test/tests/rackhd20/test_rackhd20_api_ucs.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+        self.assertEqual(len(api_data["json"]["Fabric Interconnects"]), 2,
+                         "Expected 2 Fabric Interconnects but found {0}"
+                         .format(len(api_data["json"]["Fabric Interconnects"])))
+        self.assertEqual(len(api_data["json"]["Servers"]), 7, "Expected 7 Servers but found {0}"
+                         .format(len(api_data["json"]["Servers"])))
+        self.assertEqual(len(api_data["json"]["FEX"]), 2, "Expected 2 FEX but found {0}"
+                         .format(len(api_data["json"]["FEX"])))
+        self.assertEqual(len(api_data["json"]["Chassis"]), 4, "Expected 2 Chassis but found {0}"
+                         .format(len(api_data["json"]["Chassis"])))
```
Comment (hohene) [inline test/tests/rackhd20/test_rackhd20_api_ucs.py]: design test question - do you want the test to check each of these items or exit on the first error hit?
if check each item and report errors, consider adding the errors to an errorlist, and then assert at the end if errorlist is not empty (examples of this in a couple scripts, ie. ./compute/test_rackhd20_computenode_pollers.py)

File: test/tests/rackhd20/test_rackhd20_api_ucs.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+    def test_api_20_ucs_get_rackmount(self):
+        """
+        Test the /rackmount ucs API
+        :return:
+        """
+        url = self.ucs_url_factory("rackmount")
+        api_data = fit_common.restful(url)
+        self.assertEqual(api_data['status'], 200,
+                         'Incorrect HTTP return code, expected 201, got:' + str(api_data['status']))
```
Comment (hohene) [inline test/tests/rackhd20/test_rackhd20_api_ucs.py]: change comment, expected 200 as opposed to 201

File: test/tests/rackhd20/test_rackhd20_api_ucs.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+
+    def test_api_20_ucs_get_chassis(self):
+        """
+            Test the /chassis ucs API
+            :return:
+            """
+        url = self.ucs_url_factory("chassis")
+        api_data = fit_common.restful(url)
+        self.assertEqual(api_data['status'], 200,
+                         'Incorrect HTTP return code, expected 201, got:' + str(api_data['status']))
```
Comment (hohene) [inline test/tests/rackhd20/test_rackhd20_api_ucs.py]: change comment, expected 200 as opposed to 201

File: test/tests/rackhd20/test_rackhd20_api_ucs.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+
+    def test_api_20_ucs_get_catalog(self):
+        """
+        Test the /sys ucs API
+        :return:
+        """
+        url = self.ucs_url_factory("sys")
+        api_data = fit_common.restful(url)
+        self.assertEqual(api_data['status'], 200,
+                         'Incorrect HTTP return code, expected 201, got:' + str(api_data['status']))
```
Comment (hohene) [inline test/tests/rackhd20/test_rackhd20_api_ucs.py]: change comment, expected 200 as opposed to 201

File: test/tests/rackhd20/test_rackhd20_api_ucs.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+        total_elements = 0
+        for elementTypes in api_data["json"]:
+            for element in api_data["json"][str(elementTypes)]:
+                url = self.ucs_url_factory("catalog", identifier=element["relative_path"].split("/")[-1])
+                api_data_c = fit_common.restful(url)
+                self.assertEqual(api_data['status'], 200,
+                                 'Incorrect HTTP return code, expected 201, got:' + str(api_data_c['status']))
```
Comment (hohene) [inline test/tests/rackhd20/test_rackhd20_api_ucs.py]: change comment, expected 200 as opposed to 201

**Comment by tannoa2** [comment]
@hohene 
The output of the test could be found here. It is passing....

File: test/tests/rackhd20/test_rackhd20_api_ucs.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+import fit_common
+import time
+from nosedep import depends
+import flogging
+from nose.plugins.attrib import attr
+
+logs = flogging.get_loggers()
+
+
+@attr(all=True, regression=True, smoke=True)
```
Comment (tannoa2) [inline test/tests/rackhd20/test_rackhd20_api_ucs.py]: Hi @hohene , yes this test will run as a PR Gate. I will add the attribute, thanks

**Comment by tannoa2** [comment]
@Erika, what is the best way to split this file in your opinion. Please see @geoff-reid 's comment about that

**Comment by hohene** [comment]
@tannoa2 You could start with creating a feature directory, like test/tests/ucs and move the scripts there instead of under test/tests/rackhd20. then we can split it out.

File: test/tests/rackhd20/test_rackhd20_api_ucs.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+Author(s):
+
+UCS test script that tests:
+-All the ucs service APIs
+-The Discovery workflow
+-The Catalog workflow
+
+'''
+
+import fit_common
```
Comment (hohene) [inline test/tests/rackhd20/test_rackhd20_api_ucs.py]: template was updated a bit add imports:
```
import fit_path  # NOQA: unused import
import unittest
from common import fit_common
```

File: test/tests/rackhd20/test_rackhd20_api_ucs.py [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+import time
+from nosedep import depends
+import flogging
+from nose.plugins.attrib import attr
+
+logs = flogging.get_loggers()
+
+
+@attr(all=True, regression=True, smoke=True)
+class rackhd20_api_nodes(fit_common.unittest.TestCase):
```
Comment (hohene) [inline test/tests/rackhd20/test_rackhd20_api_ucs.py]: Also, change the class name, this shows up as the test suite.
ucs_api_node  or rackhd_ucs_api_tests...something good.
remove 'fit_common.', not needed  (when you update the imports).
  ie.  just need (unittest.TestCase)

File: test/tests/rackhd20/test_rackhd20_api_ucs.py [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+
+        # Restore the nodes, obms, and catalogs to their state before the UCS discovery
+        # in order to avoid any failure in other tests
+        logs.info_1("Restoring the database to the state it was in before the UCS discovery and catalog")
+        self.restore_node_utility(catalog_workflows)
+        self.restore_obms_utility()
+
+
+if __name__ == '__main__':
+    fit_common.unittest.main()
```
Comment (hohene) [inline test/tests/rackhd20/test_rackhd20_api_ucs.py]: after updating imports, change from fit_common.unittest.main() to unittest.main()

**Comment by tannoa2** [comment]
@hohene @geoff-reid  Fixed the comment. I moved the file into /ucs directory

**Comment by hohene** [comment]
test this please

**Pull request merged**